src/Compilers/CSharp/Portable/FlowAnalysis/NullableWalker.cs

```diff
@@ -868,21 +871,47 @@ void makeNotNullMembersMaybeNull()
                     }
                 }
 
-                bool needsDefaultInitialStateForMembers()
+                ImmutableArray<Symbol> getMembersNeedingDefaultInitialState()
                 {
                     if (_hasInitialState)
                     {
-                        return false;
+                        return ImmutableArray<Symbol>.Empty;
                     }
 
                     // We don't use a default initial state for value type instance constructors without `: this()` because
                     // any usages of uninitialized fields will get definite assignment errors anyway.
-                    if (!method.HasThisConstructorInitializer(out _) && (!method.ContainingType.IsValueType || method.IsStatic))
+                    bool hasConstructorInitializer = method.HasThisConstructorInitializer(out _);
+                    if (!hasConstructorInitializer && (!method.ContainingType.IsValueType || method.IsStatic))
                     {
-                        return true;
+                        return membersToBeInitialized(method.ContainingType, includeAllMembers: true);
                     }
 
-                    return method.IncludeFieldInitializersInBody();
+                    return membersToBeInitialized(method.ContainingType, includeAllMembers: method.IncludeFieldInitializersInBody());
```

Comment:
nit: consider factoring the two calls to `membersToBeInitialized` into one (they only differ by value of `includeAllMembers`).
Also, consider leaving a PROTOTYPE marker: I assume this will need to be updated again for `SetRequiredMembers` constructors.